pkg/operatorsource/deleted.go

```diff
@@ -0,0 +1,58 @@
+package operatorsource
+
+import (
+	"context"
+
+	"github.com/operator-framework/operator-marketplace/pkg/apis/marketplace/v1alpha1"
+	"github.com/operator-framework/operator-marketplace/pkg/datastore"
+	log "github.com/sirupsen/logrus"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// NewDeletedReconciler returns a Reconciler that reconciles
+// an OperatorSource that has been marked for deletion.
+func NewDeletedReconciler(logger *log.Entry, datastore datastore.Writer, client client.Client) Reconciler {
+	return &deletedReconciler{
+		logger:    logger,
+		datastore: datastore,
+		client:    client,
+	}
+}
+
+// deletedReconciler is an implementation of Reconciler interface that
+// reconciles an OperatorSource object that has been marked for deletion.
+type deletedReconciler struct {
+	logger    *log.Entry
+	datastore datastore.Writer
+	client    client.Client
+}
+
+// Reconcile reconciles an OperatorSource object that is marked for deletion.
+//
+// in represents the original OperatorSource object received from the sdk
+// and before reconciliation has started.
+//
+// out represents the OperatorSource object after reconciliation has completed
+// and could be different from the original. The OperatorSource object received
+// (in) should be deep copied into (out) before changes are made.
+//
+// nextPhase represents the next desired phase for the given OperatorSource
+// object. If nil is returned, it implies that no phase transition is expected.
+func (r *deletedReconciler) Reconcile(ctx context.Context, in *v1alpha1.OperatorSource) (out *v1alpha1.OperatorSource, nextPhase *v1alpha1.Phase, err error) {
+	out = in
+
+	// Remove the opsrc finalizer from the object
+	out.RemoveFinalizer()
+
+	err = r.client.Update(context.TODO(), out)
```

Comment:
Is this `Update()` required given we call it in the main handler?

Reply:
We are not setting `nextPhase` because we aren't transitioning into a phase after using the delete reconciler. So either we would need to create a phase after delete reconciler which exists purely to follow this pattern and be a no-op, or we would have to use the client to update the operator source outside of the transition function (like it is currently written)

Reply:
It might be required, since you are not changing the phase the transitioner will deem as if nothing has changed. We don't do a `out.DeepEquals(in)` yet.

Reply:
Please add a comment. This is not obvious to the reader.